src/parallel/task_executor.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2025, Sirius Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "parallel/task_executor.hpp"
+
+#include "duckdb/common/helper.hpp"
+
+namespace sirius {
+namespace parallel {
+
+void ITaskExecutor::Start() {
+  bool expected = false;
+  if (!running_.compare_exchange_strong(expected, true)) {
+    return;
+  }
+  OnStart();
+  threads_.reserve(config_.num_threads);
+  for (int i = 0; i < config_.num_threads; ++i) {
+    threads_.push_back(
+      duckdb::make_uniq<WorkerThread>(duckdb::make_uniq<std::thread>(&ITaskExecutor::WorkerLoop, this, i)));
+  }
+}
+
+void ITaskExecutor::Stop() {
+  bool expected = true;
+  if (!running_.compare_exchange_strong(expected, false)) {
+    return;
+  }
+  OnStop();
+  for (auto& thread : threads_) {
+    thread->internal_thread_->join();
```

Comment:
stopping can usually be forced or non forced. I think a threadpool has a cleaner api where  instead of joining threads we focus on something thats more like a futures api where you schedule work against a threadpool and get back a future rather than controlling the threads directly

Reply:
we can change the scheduling policies by not oversbuscribing to it. at that point I do get a weird feeling about not using it so much as a thread pool and it starts to look more like a vector of threads as you proposed.